distributed/cli/dask_worker.py

```diff
@@ -386,26 +387,32 @@ def del_pid_file():
         for i in range(nprocs)
     ]
 
-    async def close_all():
-        # Unregister all workers from scheduler
-        if nanny:
-            await asyncio.gather(*[n.close(timeout=2) for n in nannies])
-
     signal_fired = False
 
-    def on_signal(signum):
+    async def _on_signal(signum):
         nonlocal signal_fired
-        signal_fired = True
-        if signum != signal.SIGINT:
+        from distributed.utils import log_errors
+
+        with log_errors():
             logger.info("Exiting on signal %d", signum)
-        asyncio.ensure_future(close_all())
+            signal_fired = True
+            if signum == signal.SIGINT:
+                logger.info("Gracefully closing worker because of SIGINT call")
+                await asyncio.gather(*[n.close_gracefully() for n in nannies])
+            logger.info("Closing workers")
+            await asyncio.gather(*[n.close() for n in nannies])
+
+    def on_signal(sig):
+        asyncio.ensure_future(_on_signal(sig))
 
     async def run():
         await asyncio.gather(*nannies)
         await asyncio.gather(*[n.finished() for n in nannies])
 
-    install_signal_handlers(loop, cleanup=on_signal)
-
+    for sig in [signal.SIGINT, signal.SIGTERM]:
+        asyncio.get_event_loop().add_signal_handler(
+            sig, functools.partial(on_signal, sig)
+        )
```

Comment:
I expect windows will not be happy with this. https://stackoverflow.com/questions/45987985/asyncio-loops-add-signal-handler-in-windows